### ISSUE TITLE
Encounter "Neo.DatabaseError.Statement.ExecutionFailed" when using WHERE in the relationship pattern

### ISSUE THREAD

**Issue by joyemang33** [neo4j/neo4j#13234]
- Neo4j version: 5.8.0-enterprise
- Operating system: Ubuntu 20.04
- API/Driver: Cypher
- **Steps to reproduce**

Hello! While testing Neo4j, I met an unexpected exception when using WHERE in the relationship pattern.
```cypher
MATCH (n1)-[r1 WHERE (r1.p18 <> r1.p16) AND (r1.p27 < r1.p7)]-(n2:(!L0|!L4)) RETURN *
```

The exception information is as follows
![image](https://github.com/neo4j/neo4j/assets/37923722/4eec94d4-e03a-4955-a7c0-807b3224ef4d)

Additional information is that this query is OK to execute in the empty graph without any syntax error. However, when executing it in my graph, the exception occurs.

Would it is possible for you to further confirm and investigate this issue? It would be highly appreciated.

The graph is created by the following Cypher statements.

[create.log](https://github.com/neo4j/neo4j/files/11883233/create.log)

Best regards,
Joye

**Comment by pontusmelke** [comment]
Thanks for reporting @joyemang33, we'll start an investigation of the issue.

Best regards,
Pontus

**Comment by pontusmelke** [comment]
@joyemang33 Hello again, looks like the issue has been adressed already. Could you test running on 5.9.0 instead?

Regards,
Pontus

**Comment by joyemang33** [comment]
Thanks for your reply! It looks like it has been fixed on 5.9.0.
